cmd/driver-registrar/main.go

```diff
@@ -50,14 +55,103 @@ const (
+	version     = "unknown"
+	// List of supported versions
+	supportedVersions = []string{"0.2.0", "0.3.0"}
 )
 
+// registrationServer is a sample plugin to work with plugin watcher
+type registrationServer struct {
+	driverName string
+	grpcServer *grpc.Server
+	endpoint   string
+	version    []string
+}
+
+// NewregistrationServer returns an initialized registrationServer instance
+func newRegistrationServer() *registrationServer {
+	return &registrationServer{}
+}
+
+// GetInfo is the RPC invoked by plugin watcher
+func (e *registrationServer) GetInfo(ctx context.Context, req *registerapi.InfoRequest) (*registerapi.PluginInfo, error) {
+	glog.Infof("Received GetInfo call: %+v", req)
+	return &registerapi.PluginInfo{
+		Type:              registerapi.CSIPlugin,
+		Name:              e.driverName,
+		Endpoint:          e.endpoint,
+		SupportedVersions: e.version,
+	}, nil
+}
+
+func (e *registrationServer) NotifyRegistrationStatus(ctx context.Context, status *registerapi.RegistrationStatus) (*registerapi.RegistrationStatusResponse, error) {
+	glog.Infof("Received NotifyRegistrationStatus call: %+v", status)
+	if !status.PluginRegistered {
+		glog.Errorf("Registration process failed with error: %+v, restarting registration container.", status.Error)
+		os.Exit(1)
+	}
+
+	return &registerapi.RegistrationStatusResponse{}, nil
+}
+
+// Serve starts Registration grpc server
+func (e *registrationServer) Serve(socketPath string) error {
+	// Remove any socket, stale or not, but fall through for other files
+	fi, err := os.Stat(socketPath)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(socketPath)
+	}
+	// Default to only user accessible socket, caller can open up later if desired
+	oldmask := unix.Umask(0077)
+
+	glog.Infof("Starting Registration Server at: %s\n", socketPath)
+	lis, err := net.Listen("unix", socketPath)
+	if err != nil {
+		return err
+	}
+	unix.Umask(oldmask)
+	glog.Infof("Registration Server started at: %s\n", socketPath)
+	e.grpcServer = grpc.NewServer()
+	// Registers kubelet plugin watcher api.
+	registerapi.RegisterRegistrationServer(e.grpcServer, e)
+
+	// Starts service
+	go func() {
+		if err := e.grpcServer.Serve(lis); err != nil {
+			glog.Errorf("Registration Server stopped serving: %v", err)
+		}
+	}()
+	return nil
+}
+
+// Called if kubernetes is configured with PluginsWatcher for driver registration
+func k8sRegistration(driverName string) {
+	stopCh := make(chan bool)
+	registrar := newRegistrationServer()
+	registrar.driverName = driverName
+	registrar.endpoint = *kubeletRegistrationPath
+	registrar.version = supportedVersions
+	glog.Infof("registrar parameters %+v", registrar)
+	registrationSocket := fmt.Sprintf("/registration/%s-reg.sock", driverName)
+	err := registrar.Serve(registrationSocket)
+	if err != nil {
+		glog.Errorf("Failed to start server with error: %+v", err)
+		os.Exit(1)
+	}
+	// Block on the channel to serve registration requests from kubelet
+	<-stopCh
```

Comment:
this will block, so it looks like the rest of the code in main does not get executed, but that is the part that labels the node. Do we not need to do that anymore if registration server is started?

Reply:
Please remember, that labeling is done by kubelet. driver-registrar labels node only if the `old` way is used.